### PR TITLE
Port storytext to use ICU BreakIterator

### PR DESCRIPTION
**Pull request by Fahad-Alsaidi**
By this commit most used storytext functions ported to use ICU BreakIterator. However, there are more to port but not used anywhere.

For testing, use check spelling  from extras menu.

### CODE REVIEW

File: scribus/text/storytext.cpp [hunk 1]
```diff
@@ -198,21 +199,22 @@ void StoryText::moveCursorLeft(bool isGrapheme)
...
 	if (isGrapheme)
 	{
-		int pos;
-		QTextBoundaryFinder boundaryFinder(QTextBoundaryFinder::Grapheme, plainText());
+		BreakIterator* it = getGraphemeIterator();
+		if (!it)
+			return;
+
+		int pos = cursorPosition();
+		it->setText(plainText().utf16());
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Since we always call `setText()` after getting the iterator, I’d move to the `get*Iterator()` method and pass `painText()` to it. This also makes sure no use site misses setting the text.

File: scribus/text/storytext.cpp [hunk 2]
```diff
@@ -198,21 +199,22 @@ void StoryText::moveCursorLeft(bool isGrapheme)
...
-			for (pos = 1; !boundaryFinder.isAtBoundary(); pos++)
-				boundaryFinder.setPosition(boundaryFinder.position() + 1);
+			pos = it->following(pos);
 		}
 		else
 		{
-			boundaryFinder.setPosition(cursorPosition() - 1);
-			for (pos = -1; !boundaryFinder.isAtBoundary(); pos--)
-				boundaryFinder.setPosition(boundaryFinder.position() - 1);
+			pos = cursorPosition();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This is redundant, you already set `pos = cursorPosition()` before the if block.

File: scribus/text/storytext.cpp [hunk 3]
```diff
@@ -252,21 +254,22 @@ void StoryText::moveCursorRight(bool isGrapheme)
...
+			return;
+
+		int pos = cursorPosition();
+		it->setText(plainText().utf16());
 		if (paragraphStyle().direction() == ParagraphStyle::RTL)
 		{
-			boundaryFinder.setPosition(cursorPosition() - 1);
-			for (pos = -1; !boundaryFinder.isAtBoundary(); pos--)
-				boundaryFinder.setPosition(boundaryFinder.position() - 1);
+			pos = cursorPosition();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Same here.

Comment (khaledhosny) [inline scribus/text/storytext.cpp]: After removing this line, you can remove the curly brackets around the blocks as they will be redundant.

File: scribus/text/storytext.cpp [hunk 4]
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
...
-	if (pos < len)
+	BreakIterator* it = getWordIterator();
+	if (!it)
 		return pos;
-	else
-		return len;
+
+	it->setText(plainText().utf16());
+	int len = length();
+	int nxt = it->following(pos);
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Just make is `pos = …` and return pos, instead of the ugly variable name.

File: scribus/text/storytext.cpp [hunk 5]
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
...
-	//if we didn't get to the end, return current position, otherwise return the end.
-	if (pos < len)
+	BreakIterator* it = getWordIterator();
+	if (!it)
 		return pos;
-	else
-		return len;
+
+	it->setText(plainText().utf16());
+	int len = length();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This variable is unused, are you sure the new code matches the behaviour of the old one?

File: scribus/text/storytext.cpp [hunk 6]
```diff
@@ -1540,49 +1537,53 @@ int StoryText::prevWord(int pos)
 	return wordBoundaries.indexOf(text(pos)) < 0 ? pos + 1 : pos;
 }
 
-int StoryText::endOfWord(int pos) const
+int StoryText::endOfWord(int pos)
 {
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Why was the `const` removed?

File: scribus/text/storytext.cpp [hunk 7]
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
...
-	else
-		return len;
+
+	it->setText(plainText().utf16());
+	int len = length();
+	int nxt = it->following(pos);
+	return nxt;
 }
 
 int StoryText::prevWord(int pos)
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This should be ported as well, to match the method above.

File: scribus/text/storytext.h [hunk 8]
```diff
@@ -274,12 +274,15 @@ public slots:
...
-
 private:
 	ScribusDoc * m_doc; 
 	int m_selFirst, m_selLast;
+	static BreakIterator* m_graphemeIterator;
 	static BreakIterator* m_wordIterator;
+	static BreakIterator* m_sentenceIterator;
+	BreakIterator* getGraphemeIterator();
 	BreakIterator* getWordIterator();
+	BreakIterator* getSentenceIterator();
```
Comment (khaledhosny) [inline scribus/text/storytext.h]: These methods should be static as well, I missed that for `getWordIterator()`.

File: scribus/text/storytext.cpp (hunk 7 above)
Comment (Fahad-Alsaidi) [inline scribus/text/storytext.cpp]: This function not used anywhere. So no point to port it.

File: scribus/text/storytext.cpp (hunk 5 above)
Comment (Fahad-Alsaidi) [inline scribus/text/storytext.cpp]: Yes I sure. You can test by using Spelling check from extras menu.

Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Remove the variable then.

File: scribus/text/storytext.cpp (hunk 7 above)
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Either port it for completeness sake, since it might get be used in the future and it will be wrong then, or completely remove it. I prefer keeping it for API parity, it does not hurt much.

File: scribus/text/storytext.cpp (hunk 1 above)
Comment (Fahad-Alsaidi) [inline scribus/text/storytext.cpp]: I tried this but cases a lot of problem. Iterator starts mix text and never work with CJK. I think the problem that we use static vars which is unsafe thread. So better leave it as it unless there is somebody willing to dig ICU source code to find the problem.

**Comment by Fahad-Alsaidi** [comment]
Updated. Please notice that I return endOfWord back the const and make all similar functions const. Also I port prevWord & prevSentence to ICU even they are not used anywhere. I also modified to adopt the new change.

File: scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp [hunk 9]
```diff
@@ -181,7 +181,7 @@ bool HunspellPluginImpl::parseTextFrame(StoryText *iText)
 			wf.replacements = hspellerMap[wordLang]->suggest(word);
 			wordsToCorrect.append(wf);
 		}
-		currPos = iText->nextWord(wordStart);
+		currPos = iText->endOfWord(wordStart);
```
Comment (khaledhosny) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: Why was this changed?

File: scribus/text/storytext.cpp [hunk 10]
```diff
@@ -167,7 +168,7 @@ int StoryText::cursorPosition() const
 	return d->cursorPosition;
 }
 
-void StoryText::setCursorPosition(int pos, bool relative)
+void StoryText::setCursorPosition(int pos, bool relative) const
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This changes the cursor position, so it shouldn’t be const. Only methods that changes nothing in the class should be const.

File: scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp (hunk 9 above)
Comment (Fahad-Alsaidi) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: because nextWord will give us word+space and we don't want it.

File: scribus/text/storytext.cpp [hunk 11]
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
...
-	if (pos < len)
+	BreakIterator* it = getWordIterator();
+	if (!it)
 		return pos;
-	else
-		return len;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	pos = it->next();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp:1519]: Why is the call to `next()` here?

File: scribus/text/storytext.cpp [hunk 12]
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
...
-	pos = qMax(0, pos-1);
-	while (pos > 0 && wordBoundaries.indexOf(text(pos)) < 0)
-		--pos;
-	return wordBoundaries.indexOf(text(pos)) < 0 ? pos + 1 : pos;
+	BreakIterator* it = getWordIterator();
+	if (!it)
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This should `preceding()`, shouldn’t it?

File: scribus/text/storytext.cpp [hunk 13]
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
...
-	while (pos < len && sentenceBoundaries.indexOf(text(pos)) >= 0)
-		++pos;
-	return pos;
+	BreakIterator* it = getSentenceIterator();
+	if (!it)
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	int next = it->next();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Just make it `pos = it->next()` and return `pos`, no need for a new variable. Similar to code above.

File: scribus/text/storytext.cpp [hunk 14]
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
...
-	while (pos > 0 && sentenceBoundaries.indexOf(text(pos-1)) < 0)
-		--pos;
-	return pos;
+	BreakIterator* it = getSentenceIterator();
+	if (!it)
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	int prev = it->previous();
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: Same here.

File: scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp (hunk 9 above)
Comment (khaledhosny) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: But then the new `nextWord()` is wrong, it should match exactly the behaviour of the old one.

Comment (Fahad-Alsaidi) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: In that case I will make nextWord call endOfWord because they will be same just duplicated code. Is that right?

File: scribus/text/storytext.cpp (hunk 11 above)
Comment (Fahad-Alsaidi) [inline scribus/text/storytext.cpp:1519]: to get the begging of the next word after the space.

**Comment by Fahad-Alsaidi** [comment]
updated!

File: scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp (hunk 9 above)
Comment (khaledhosny) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: If this is the only place this was calling `nextWord()`, then keep this change and just kill `nextWord()` and `prevWord()` since both will be unused now.

File: scribus/text/storytext.cpp [hunk 15]
```diff
@@ -1579,9 +1570,13 @@ int StoryText::prevSentence(int pos)
...
 		return 0;
-	//while not on a sentence boundary, keep moving backward
-	while (pos > 0 && sentenceBoundaries.indexOf(text(pos-1)) < 0)
-		--pos;
+	BreakIterator* it = getSentenceIterator();
+	if (!it)
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
```
Comment (khaledhosny) [inline scribus/text/storytext.cpp]: This should be `preceding()` as well.

File: scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp (hunk 9 above)
Comment (Fahad-Alsaidi) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: As we agreed on https://github.com/HOST-Oman/scribus/issues/166 I will not remove any unused code.

Comment (khaledhosny) [inline scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:184]: Yes, I was under the assumption that we are not touching that unused code at all, but if we need to modify it because it is broken then we are better of just killing it.

**Comment by Fahad-Alsaidi** [comment]
updated!

**Comment by andreas-vox** [comment]
Question: what are the advantages of the ICU break iterator over Qt's QTextBoundaryFinder?

**Comment by khaledhosny** [comment]
@Fahad-Alsaidi reported that the word iterator was not working for some languages, also ICU BreakIterator is more widely used and tested, and gets updated regularly for new Unicode versions. Qt itself is going to use ICU at some point.

**Comment by Fahad-Alsaidi** [comment]
@andreas-vox  For your reference please see these bugs:
1- [QTBUG-53034 ](https://bugreports.qt.io/browse/QTBUG-53034) QTextBoundaryFinder::Word doesn't work in CJK
2- [QTBUG-29930](https://bugreports.qt.io/browse/QTBUG-29930) Unicode script itemization code needs to be re-worked
3- [QTBUG-42078](https://bugreports.qt.io/browse/QTBUG-42078) Support Lao Word Boundary Detection and Line Breaking
4- [QTBUG-33763](https://bugreports.qt.io/browse/QTBUG-33763) Update Unicode database and algorithms up to 8.0

**Comment by andreas-vox** [comment]
Thanks for the links and your good work on CTL!

So currently Harfbuzz + ICU are state of the art?

**Comment by khaledhosny** [comment]
Yes AFAIK, that is pretty much what everyone is using or migrating to (Firefox, Chrome, LibreOffice, Java etc.) Even InDesign uses parts of ICU (but not HarfBuzz yet AFAIK, we have got a big head start over them in this area 😃).
